### ISSUE TITLE
Dynamic schema package access for plugins (aka. allow Nexus Schema to be wrapped/embedded by other tools)

### ISSUE THREAD

**Issue by jasonkuhrt** [graphql-nexus/schema#413]
While building the framework we have discovered that the assumption that Nexus Schema has of it being a top-level dependency and thus something that plugins should put in their peer-dependencies goes against being embedded/wrapped by another tool (in our case the framework).

Nexus Framework bundles Nexus Schema and Nexus Framework plugins may bundle Nexus Schema plugins. This creates a situation where the import of `@nexus/schema` within a schema plugin is based on package _hoisting_. This is problematic and we want to as a design goal get to zero-dependence on package hoisting behaviour.

The background on how we got here can be found in writing in these two comments:

- https://github.com/graphql-nexus/nexus/issues/514#issuecomment-604668904
- https://github.com/graphql-nexus/nexus/issues/514#issuecomment-605105455

Although not seen here are multi-hour design discusses between @weakky and I preceding the comments.

We currently see the solution being roughly as follows:

- Nexus Schema should get rid of the plugin class and wrapper-function

- Nexus Schema should only statically export plugin _types_
- Nexus Schema should pass plugins _as an arg_ the Nexus Schema package (or subset, as we deem correct) to give the plugin access to the things it would have normally imported statically
- Nexus Schema plugins should pin to Nexus Schema as a dev dep
- Nexus Schema plugins should import a plugin type from Nexus Schema
- Nexus Schema plugins should _not_ have Nexus Schema as a peer dep
- Nexus Schema plugins should pass version of Nexus they are compatible with as runtime string that Nexus Nexus can validate (achieving similar to what unmet peer dep warnings provide, precedent/ref for this: Fastify)
- ditto the above for graphql package if we deem necessary
- via the dynamic access to Nexus, plugins can still get a hold of `completeValue`, so we don't need to talk about removing it here.

**Comment by Nayni** [comment]
> Nexus Schema plugins should not have Nexus Schema as a peer dep

From the Nexus framework POV this may be something that solves some of your problems with package hoisting, but I think it's a rather poor solution in stand-alone usages of `@nexus/schema` because my package manager will not give me any feedback that I've installed a plugin but don't have the schema component yet (or that I've installed a plugin which is not meeting its `peerDependency` version range, the latter is "solved" but I will only know this at runtime)

I might've missed a lot of valid discussion points and I'll be honest I've only skimmed through the threads that were linked so I might be missing some vital pieces here, but I feel like not using npm's built-in way of specifying `peerDependencies` is weird.

**Comment by jasonkuhrt** [comment]
> what do you see happen with the plugins that are now packaged together with @nexus/schema? Would you move them into specific plugin packages too?

Yeah I could see them as standalone packages in the future. On how to name them, maybe `nexus-schema-plugin-*`.

> my package manager will not give me any feedback that I've installed a plugin but don't have the schema component yet (or that I've installed a plugin which is not meeting its peerDependency version range, the latter is "solved" but I will only know this at runtime)

We could probably do better than the feedback provided by peer dependency warnings via post install hooks.

The number of peer dep warnings that users have become habituated to ignoring seriously puts into question the usability of the system I think.

There are existing plugin systems based on runtime checks, [at least Fastify](https://github.com/fastify/fastify/blob/master/docs/Plugins.md#create-a-plugin).

> Creating a plugin is very easy, you just need to create a function that takes three parameters, the fastify instance, an options object and the done callback.

They have a plugin helper as a [separate package](https://github.com/fastify/fastify-plugin) offering the runtime version compat checking

```ts
const fp = require('fastify-plugin')

module.exports = fp(function (fastify, opts, next) {
  // your plugin code
  next()
}, { fastify: '1.x' })
```

Having plugins receive their "host" rather than statically importing their host makes a whole lot of sense I think.

But anyways, Nexus has an unorthadox two-tier plugin architecture and it fundamentally does not work with the peer dependency system which is designed for plugin or like systems where the specified peer dep is expected to be a top-level project dependency.

The design of Nexus Schema today makes it so that it cannot be embedded by other tools without the unwanted effect of its plugin system relying on acquiring it via package hoisting.

**Comment by Nayni** [comment]
Ok, those are some fair points @jasonkuhrt 

I actually like the Fastify approach and I would promote having a `nexus-schema-plugin` core package which can provide the helpers to make a plugin quickly as I believe those factory functions are very valuable, especially in the world of TypeScript where such functions can infer generic arguments instead of having to type them out manually.

I also agree that `peerDependency` warnings are not always the best gate of defense, but I'm always in favor of using existing systems instead of creating new ones, although I can start seeing the points that makes your proposal logical.

Although I'm mainly a consumer of `@nexus/schema` only for now, would there be a situation where a framework plugin has a dependency on a schema plugin? (to enhance it with extra behavior or default config for easier use)

**Comment by jasonkuhrt** [comment]
> would there be a situation where a framework plugin has a dependency on a schema plugin?

Yeah `nexus-plugin-prisma` for example depends on `nexus-prisma` (to be renamed `nexus-schema-plugin-prisma` probably).

But even if one installs a schema plugin into a framework project ([`schema.use`](https://www.nexusjs.org/#/api/modules/main/exports/schema?id=use)) there is still a hoisting problem, fixed only if the project explicitly depends upon `@nexus/schema`.

While we haven't been bitten by any hoisting problems yet, as the community grows and plugins increase in number, I do feel the cracks may begin to show. 🤞

**Comment by jasonkuhrt** [comment]
Another example of a plugin system that gets is host package dynamically rather than through a static import from a dep: https://github.com/microsoft/TypeScript/wiki/Writing-a-Language-Service-Plugin

**Comment by jasonkuhrt** [comment]
Another example of a plugin system that gets its host package features dynamically rather than through a static import from a dep:

https://rollupjs.org/guide/en/#plugin-development

**Comment by jasonkuhrt** [comment]
Recent review with @Weakky about why this issue is fundamentally required IF we want to avoid depending on hoisting.

---


## Question: why isn't making `graphql` a `nexus` peer dep enough to NOT have to break the "@nexus/schema" plugin interface?


### Example Why

Imagine a Nexus Schema plugin


```json
name: "foobar"
peerDependencies: {
  "@nexus/schema": "..."
}
```

```ts
import { plugin } from '@nexus/schema"  // <---------- peer dep

export default plugin(...)
```
  

### So what is the Problem?


How does the Nexus Schema plugin avoid depending on hoisting of `@nexus/schema` in a framework context?

### Solution Summaries

#### Universe A

```
- Rely on hoisting                                <-- works if one version in tree
- Yarn resolutions                                <-- works if multiple versions in tree + using yarn

#### Universe B

- User installs "@nexus/schema" in their project  <-- works if we change our architeture
```

### Hoisting Violation Scenarios (e.g. @nexus/schema)

- "bad" library transiently depends on `@nexus/schema`
- "bad" library installed by user depends on `@nexus/schema`
- user installs incompatible version of `@nexus/schema`
- yarn workspace? (no, seems fine)
  two workspaces, each with a different version of `@nexus/schema`
  ```
  Layout 1
  --------
  /node_modules
    /nexus-plugin-prisma     <-- require('@nexus/schema') ???
    /foo@1
  /packages
    /a
      /@nexus/schema@12
      /foo                   ^^^
      /nexus-plugin-prisma   ^^^
    /b
      /@nexus/schema@11
      /foo                   ^^^
      /nexus-plugin-prisma   ^^^

  Layout 2
  --------
  /node_modules
    /foo
    /@nexus/schema
    /nexus-plugin-prisma     
  /packages
    /a
      /@nexus/schema@11      ^^^
      /foo@1                 ^^^
      /nexus-plugin-prisma   ^^^
    /b
      /@nexus/schema@11      ^^^
      /foo@1                 ^^^
      /nexus-plugin-prisma   ^^^
  ```

### The hosisting effect "visually"

```
(without hoisting)

/nexus
  /@nexus/schema
  /graphql
/nexus-plugin-prisma
  /@prisma/cli
  /@prisma/client
  /@prisma/sdk

(with hoisting)

/@nexus/schema
/graphql
/nexus
/nexus-plugin-prisma
/@prisma/cli
/@prisma/client
/@prisma/sdk
```
